Sources/Machine/DensityMatrices/abstract_density_matrix.cc

```diff
@@ -0,0 +1,200 @@
+          *doubled_graph, physical_hilbert->LocalStates())),
+      hilbert_physical_(physical_hilbert),
+      graph_doubled_(std::move(doubled_graph)) {}
+
+AbstractDensityMatrix::AbstractDensityMatrix(
+    std::shared_ptr<const AbstractHilbert> hilbert)
+    : AbstractDensityMatrix(hilbert, DoubledGraph(hilbert->GetGraph())) {}
+
+std::shared_ptr<const AbstractHilbert>
+AbstractDensityMatrix::GetHilbertPhysicalShared() const {
+  return hilbert_physical_;
+}
+
+const AbstractHilbert &AbstractDensityMatrix::GetHilbertPhysical() const
+    noexcept {
+  return *hilbert_physical_;
+}
+
+void AbstractDensityMatrix::InitLookup(VisibleConstType v, LookupType &lt) {
+  // split into first half and second half
+  VisibleConstType v_r = v.head(GetHilbertPhysical().Size());
+  VisibleConstType v_c = v.tail(GetHilbertPhysical().Size());
+
+  return InitLookup(v_r, v_c, lt);
+}
+
+void AbstractDensityMatrix::UpdateLookup(VisibleConstType v,
+                                         const std::vector<int> &tochange,
+                                         const std::vector<double> &newconf,
+                                         LookupType &lt) {
+  std::vector<int> tochange_r, tochange_c;
+  std::vector<double> newconf_r, newconf_c;
+
+  if (tochange.size() != 0) {
+    for (std::size_t s = 0; s < tochange.size(); s++) {
+      const int sf = tochange[s];
+      if (sf < Nvisible()) {
+        tochange_r.push_back(sf);
+        newconf_r.push_back(newconf[s]);
+      } else {
+        tochange_c.push_back(sf - Nvisible());
+        newconf_c.push_back(newconf[s]);
+      }
+    }
+  }
+
+  // split into first half and second half
+  VisibleConstType v_r = v.head(GetHilbertPhysical().Size());
+  VisibleConstType v_c = v.tail(GetHilbertPhysical().Size());
+  return UpdateLookup(v_r, v_c, tochange_r, tochange_c, newconf_r, newconf_c,
+                      lt);
+}
+
+VectorType AbstractDensityMatrix::DerLog(VisibleConstType v) {
+  return DerLog(v.head(GetHilbertPhysical().Size()),
+                v.tail(GetHilbertPhysical().Size()));
+}
+
+VectorType AbstractDensityMatrix::DerLog(VisibleConstType v,
+                                         const LookupType &lt) {
+  return DerLog(v.head(GetHilbertPhysical().Size()),
+                v.tail(GetHilbertPhysical().Size()), lt);
+}
+
+Complex AbstractDensityMatrix::LogVal(VisibleConstType v) {
+  return LogVal(v.head(GetHilbertPhysical().Size()),
+                v.tail(GetHilbertPhysical().Size()));
+}
+
+Complex AbstractDensityMatrix::LogVal(VisibleConstType v,
+                                      const LookupType &lt) {
+  return LogVal(v.head(GetHilbertPhysical().Size()),
+                v.tail(GetHilbertPhysical().Size()), lt);
+}
+
+VectorType AbstractDensityMatrix::LogValDiff(
+    VisibleConstType v, const std::vector<std::vector<int>> &tochange,
+    const std::vector<std::vector<double>> &newconf) {
+  VisibleConstType v_r = v.head(GetHilbertPhysical().Size());
```

Comment:
Can we move this part of code in a common function that is also used by other functions (see the one below for example) ?